update-engine/src/buffer.rs

```diff
@@ -0,0 +1,913 @@
+        let mut step_events = self.event_store.high_priority.clone();
+        let mut progress_events = Vec::new();
+        for value in self.event_store.map.values() {
+            step_events.extend(value.low_priority.iter().cloned());
+            progress_events.push(value.progress_event.clone());
+        }
+
+        // Sort events by their total duration. We assume that total_elapsed is
+        // monotonically increasing. We should come up with a better sort
+        // algorithm in the future, based on the actual event kinds.
+        step_events.sort_by(|a, b| a.total_elapsed.cmp(&b.total_elapsed));
+        progress_events.sort_by(|a, b| a.total_elapsed.cmp(&b.total_elapsed));
+
+        EventReport { step_events, progress_events }
+    }
+
+    /// Clears events returned by the last call to [`Self::generate_report`].
+    ///
+    /// Call this after successfully sending a report over the wire.
+    pub fn on_report_processed(&mut self) {
+        self.event_store.clear_step_events();
+        // Do *not* clear progress events -- we always want to keep returning at
+        // least one of them, and have them be cleared in the end once a step or
+        // execution is complete.
+    }
+
+    pub fn add_progress_event(&mut self, event: ProgressEvent<S>) {
+        self.event_store.handle_progress_event(event);
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+enum CompletedIndex {
+    LastSeen(usize),
+    // Terminal is greater than any LastSeen.
+    Terminal,
+}
+
+#[derive_where(Clone, Debug, Default)]
+struct EventStore<S: StepSpec> {
+    // A tree where edges are from parent event keys to child nested event keys.
+    //
+    // While petgraph seems like overkill at first, it results in really
+    // straightforward algorithms below compared to alternatives like storing
+    // trees using Box pointers.
+    event_tree: DiGraphMap<EventKey, ()>,
+    high_priority: Vec<StepEvent<S>>,
+    map: HashMap<EventKey, EventMapValue<S>>,
+    last_completed_index: HashMap<ExecutionId, CompletedIndex>,
+}
+
+impl<S: StepSpec> EventStore<S> {
+    /// Handles a step event.
+    fn handle_step_event(
+        &mut self,
+        event: StepEvent<S>,
+        max_low_priority: usize,
+    ) {
+        if matches!(event.kind, StepEventKind::Unknown) {
+            // Ignore unknown events.
+            return;
+        }
+
+        if let Some(key) = self.recurse_for_step_event(&event, None) {
+            match self.map.entry(key) {
+                hash_map::Entry::Occupied(mut entry) => {
+                    let value = entry.get_mut();
+                    let progress_event = event.progress_event();
+                    if event.kind.priority() == StepEventPriority::High {
+                        // Handle duplicate events.
+                        if self.high_priority.last() != Some(&event) {
+                            self.high_priority.push(event);
+                        }
```

Comment:
This isn't correct -- if we want to be resilient to duplicate events, we need to search through all earlier events (and ensure `high_priority` is in sorted order).

Reply:
Done. This simplifies the algorithm quite a bit!